source/RoslynAnalyzers/PersistenceAnalyzers.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers;
+
+using static Descriptors;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class PersistenceAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+        DontUseIDocumentStoreOfEvent);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        if (!Debugger.IsAttached) context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(DontUse_IDocumentStoreOfEvent, SyntaxKind.FieldDeclaration, SyntaxKind.PropertyDeclaration, SyntaxKind.ConstructorDeclaration);
+    }
+
+    // this flags declarations of fields or properties referencing IDocumentStore<Event, string>,
+    // as well as constructor parameters (assuming someone is importing it from Autofac)
+    // it doesn't flag on other things to reduce the effort required to check for this.
+    void DontUse_IDocumentStoreOfEvent(SyntaxNodeAnalysisContext context)
+    {
+        static void CheckForIDocumentStoreOfEvent(SyntaxNodeAnalysisContext context, TypeSyntax typeToCheck)
+        {
+            GenericNameSyntax genericNameSyntax;
+            switch (typeToCheck)
+            {
+                case GenericNameSyntax gns:
+                    genericNameSyntax = gns;
+                    break;
+                case NullableTypeSyntax { ElementType: GenericNameSyntax gns }:
+                    genericNameSyntax = gns;
+                    break;
+                default:
+                    return;
+            }
+
+            if (genericNameSyntax.Identifier.Text == "IDocumentStore" && genericNameSyntax.TypeArgumentList.Arguments.Count == 2)
+            {
```

Comment:
Specific coupling like this makes me feel like we should eventually try and pull these analyzers into the server solution itself. But a problem for later.

Reply:
Yep. That'd actually be pretty easy to do, it'd make this whole process easier.
I was going to ask you about who I should discuss that with and the best way to move forward there.

Reply:
I think Rob Wagner started this repo, so you could chat with him about it to form some opinions. The thing I wonder about is are there analyzers in here we use for other repos? So do we need a split of Server-specific analysers like this new one, and more general-purpose ones we like in all of our solutions?

Reply:
Good points, I'll discuss this with Rob and move it forward as a low-priority fill-in task

Reply:
Ha, great minds think alike! We were discussing this in standup this morning: https://octopusdeploy.slack.com/archives/C02N4QE2U3D/p1672872922444639

@borland We'll wait to hear about the results from your chat with Rob. Please let us know if we can help. If we agree that Server-specific analyzers belong in the Server solution, we're happy to do that work in CPP to help setup the pattern